### DISCUSSION TITLE
HttpClient: reading response stream while streaming response chunk by chunk makes next chunk the last one

### DISCUSSION THREAD

**Discussion by karrakoliko** [symfony/symfony#48749] [Q&A] [answered]
### Symfony version(s) affected

5.4.16

### Description

We expect to be able to read first N bytes of ```Symfony\Contracts\HttpClient\ResponseInterface``` while streaming it chunk by chunk using ```foreach ($httpClient->stream($response) as $chunk)```.

Currently, if we try to read it - the next chunk becomes the last (see how to reproduce section, i have added phpunit test to show the behavior).

### How to reproduce

```php
<?php

use http\Client\Response;
use Symfony\Component\HttpClient\MockHttpClient;
use Symfony\Component\HttpClient\Response\MockResponse;
use Symfony\Component\HttpClient\Response\StreamWrapper;

class SymfonyMockResponseTest extends \PHPUnit\Framework\TestCase
{

    /**
     * When we read response stream chunk by chunk, we expect to have ability to read response stream.
     * Now, if we read at least one byte of response stream, or seek it - the next chunk becomes last one
     *
     * Real world example: we read rss feed chunk by chunk.
     * At some chunk we need get access to first N bytes of response body, and then continue.
     * If we do it - we can't read next chunk, as it becomes last.
     *
     * @todo: i'm not sure if it only happens to MockResponse/MockHttpClient.
     *
     * @dataProvider streamReadProvider
     * @param bool $responseStreamRead
     * @return void
     * @throws \Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface
     */
    public function testReadingResponseStreamDoesNotMakeNextChunkTheLastOne(bool $responseStreamRead)
    {
        $totalChunks = 100;

        $responseMock = new MockResponse($this->simulateNChunks($totalChunks - 1));
        $clientMock = new MockHttpClient($responseMock);

        $chunkN = 0;
        $lastChunkN = null;

        $response = $clientMock->request('GET','https://example.com');

        foreach ($clientMock->stream($response) as $chunk) {

            if ($chunk->isLast()) {
                $lastChunkN = $chunkN;
            }

            if($responseStreamRead){

                if (!$chunk->isLast()) {
                    $stream = StreamWrapper::createResource($response);
                    $pos = ftell($stream);

                    // both fread/fseek cause next chunk to be the last one
                    fread($stream, 1);
                    fseek($stream,$pos);
                }

            }

            $chunkN++;

        }

        $this->assertEquals($totalChunks, $lastChunkN);
    }

    protected function simulateNChunks(int $n)
    {
        while($n > 0){
            yield 'chunk#' . $n . ';';
            $n--;
        }

    }

    public function streamReadProvider()
    {
        return [
            'dont read response stream while iterating chunks' => [false],
            'read response stream while iterating chunks' => [true],
        ];
    }

}
```

**Comment by nicolas-grekas** [comment]
I don't know how this will behave with real responses.
But the code looks terrible to me. If you want the first bytes, just store them when they're yielded. :shrug: 
PR welcome if you figure out a way to improve this. I'd also be fine closing personally :)

**Comment by karrakoliko** [comment]
@nicolas-grekas my expectation is that we can use both chunk-based and stream-based access to response body. Simultaneously: if we read response body chunk-by-chunk, we can access to response body stream (while reading chunk), and vice versa (we can chunk response body stream).

Obviously, this expectation is wrong.

Could you clarify: internally, when we send request, httpClient gets chunks of response body, and then creates stream of them, or vice versa (creates stream, and then converts to chunks)?

**Comment by nicolas-grekas** [comment, marked as answer]
So, I spent some time on this and the behavior is the following:
- the stream() methods yield *when there is activity on the network*
- the stream resource consumes all already received packets waiting in the network stack.

This means that when you use the stream while streaming, you will get less chunk objects, because some will be consumed by the call to fread().
When using MockHttpClient, because there is no real network, using fread will read the full body. Then, when resuming the foreach, only the last chunk will remain.
This means that the behavior you're observing is correct: you effectively consumed the stream. If you want to get the corresponding content, you can seek+read the resource.
There is a competition between both APIs (but both react to network events, and that's the only behavior that makes sense when the response body is not buffered, which both APIs support.)
